Does not build with LibreSSL 2.8.0

Hi,

LibreSSL 2.8.0 is released but not supported by this crate. Bumping the version numbers in the build scripts is not enough to compile it: There have been some `const` changes in LibreSSL that break tests and most likely the build as well.

This also repeats #939 and breaks the crate and any dependents on OpenBSD 6.4-beta which comes with LibreSSL 2.8.0.

Is there any work on supporting LibreSSL 2.8.0 yet or should we look into it?

Reyk

```make
   Compiling systest v0.1.0 (file:///rust/rust-openssl/systest)
error: failed to run custom build command for `systest v0.1.0 (file:///rust/rust-openssl/systest)`
process didn't exit successfully: `/rust/rust-openssl/target/debug/build/systest-d7f8964f8ffe8d09/build-script-build` (exit code: 101)
--- stdout
-----------------------------------------
cargo:rerun-if-changed=../openssl-sys/src/lib.rs
cargo:rerun-if-changed=../openssl-sys/src/libressl/mod.rs
cargo:rerun-if-changed=../openssl-sys/src/libressl/v251.rs
cargo:rerun-if-changed=../openssl-sys/src/libressl/v273.rs
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-openbsd")
HOST = Some("x86_64-unknown-openbsd")
CC_x86_64-unknown-openbsd = None
CC_x86_64_unknown_openbsd = None
HOST_CC = None
CC = None
HOST = Some("x86_64-unknown-openbsd")
HOST = Some("x86_64-unknown-openbsd")
CFLAGS_x86_64-unknown-openbsd = None
CFLAGS_x86_64_unknown_openbsd = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
HOST = Some("x86_64-unknown-openbsd")
CFLAGS_x86_64-unknown-openbsd = None
CFLAGS_x86_64_unknown_openbsd = None
HOST_CFLAGS = None
CFLAGS = None
HOST = Some("x86_64-unknown-openbsd")
CFLAGS_x86_64-unknown-openbsd = None
CFLAGS_x86_64_unknown_openbsd = None
HOST_CFLAGS = None
CFLAGS = None
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-o" "/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.o" "-c" "/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c"
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:1461:28: error: returning 'const BIO_METHOD **' (aka 'const struct bio_method_st **') from a function with result type 'BIO_METHOD **' (aka 'struct bio_method_st **') discards qualifiers in nested pointer types [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
cargo:warning=                    return &b->method;
cargo:warning=                           ^~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2776:24: error: incompatible pointer types returning 'BIO *(const BIO_METHOD *)' (aka 'struct bio_st *(const struct bio_method_st *)') from a function with result type 'BIO *(*)(BIO_METHOD *)' (aka 'struct bio_st *(*)(struct bio_method_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return BIO_new;
cargo:warning=                       ^~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2781:24: error: incompatible pointer types returning 'const BIO_METHOD *(void)' (aka 'const struct bio_method_st *(void)') from a function with result type 'BIO_METHOD *(*)(void)' (aka 'struct bio_method_st *(*)(void)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return BIO_s_file;
cargo:warning=                       ^~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2786:24: error: incompatible pointer types returning 'const BIO_METHOD *(void)' (aka 'const struct bio_method_st *(void)') from a function with result type 'BIO_METHOD *(*)(void)' (aka 'struct bio_method_st *(*)(void)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return BIO_s_mem;
cargo:warning=                       ^~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2871:24: error: incompatible pointer types returning 'PKCS12 *(const char *, const char *, EVP_PKEY *, X509 *, struct stack_st_X509 *, int, int, int, int, int)' (aka 'PKCS12 *(const char *, const char *, struct evp_pkey_st *, struct x509_st *, struct stack_st_X509 *, int, int, int, int, int)') from a function with result type 'PKCS12 *(*)(char *, char *, EVP_PKEY *, X509 *, struct stack_st_X509 *, int, int, int, int, int)' (aka 'PKCS12 *(*)(char *, char *, struct evp_pkey_st *, struct x509_st *, struct stack_st_X509 *, int, int, int, int, int)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return PKCS12_create;
cargo:warning=                       ^~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2936:24: error: incompatible pointer types returning 'const char *(const SSL_CIPHER *)' (aka 'const char *(const struct ssl_cipher_st *)') from a function with result type 'char *(*)(const SSL_CIPHER *)' (aka 'char *(*)(const struct ssl_cipher_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return SSL_CIPHER_get_version;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2951:24: error: incompatible pointer types returning 'void (SSL_CTX *, SSL_SESSION *(*)(struct ssl_st *, const unsigned char *, int, int *))' (aka 'void (struct ssl_ctx_st *, struct ssl_session_st *(*)(struct ssl_st *, const unsigned char *, int, int *))') from a function with result type 'void (*)(SSL_CTX *, SSL_SESSION *(*)(SSL *, unsigned char *, int, int *))' (aka 'void (*)(struct ssl_ctx_st *, struct ssl_session_st *(*)(struct ssl_st *, unsigned char *, int, int *))') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return SSL_CTX_sess_set_get_cb;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2976:24: error: incompatible pointer types returning 'X509_NAME *(const X509 *)' (aka 'struct X509_name_st *(const struct x509_st *)') from a function with result type 'X509_NAME *(*)(X509 *)' (aka 'struct X509_name_st *(*)(struct x509_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_get_subject_name;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2981:24: error: incompatible pointer types returning 'X509_NAME *(const X509 *)' (aka 'struct X509_name_st *(const struct x509_st *)') from a function with result type 'X509_NAME *(*)(X509 *)' (aka 'struct X509_name_st *(*)(struct x509_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_get_issuer_name;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:2996:24: error: incompatible pointer types returning 'void *(const X509 *, int, int *, int *)' (aka 'void *(const struct x509_st *, int, int *, int *)') from a function with result type 'void *(*)(X509 *, int, int *, int *)' (aka 'void *(*)(struct x509_st *, int, int *, int *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_get_ext_d2i;
cargo:warning=                       ^~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3001:24: error: incompatible pointer types returning 'int (X509_NAME *, int, int, const unsigned char *, int, int, int)' (aka 'int (struct X509_name_st *, int, int, const unsigned char *, int, int, int)') from a function with result type 'int (*)(X509_NAME *, int, int, unsigned char *, int, int, int)' (aka 'int (*)(struct X509_name_st *, int, int, unsigned char *, int, int, int)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_NAME_add_entry_by_NID;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3006:24: error: incompatible pointer types returning 'int (const X509_NAME *)' (aka 'int (const struct X509_name_st *)') from a function with result type 'int (*)(X509_NAME *)' (aka 'int (*)(struct X509_name_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_NAME_entry_count;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3011:24: error: incompatible pointer types returning 'X509_NAME_ENTRY *(const X509_NAME *, int)' (aka 'struct X509_name_entry_st *(const struct X509_name_st *, int)') from a function with result type 'X509_NAME_ENTRY *(*)(X509_NAME *, int)' (aka 'struct X509_name_entry_st *(*)(struct X509_name_st *, int)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_NAME_get_entry;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3016:24: error: incompatible pointer types returning 'ASN1_STRING *(const X509_NAME_ENTRY *)' (aka 'struct asn1_string_st *(const struct X509_name_entry_st *)') from a function with result type 'ASN1_STRING *(*)(X509_NAME_ENTRY *)' (aka 'struct asn1_string_st *(*)(struct X509_name_entry_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_NAME_ENTRY_get_data;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3021:24: error: incompatible pointer types returning 'ASN1_OBJECT *(const X509_NAME_ENTRY *)' (aka 'struct asn1_object_st *(const struct X509_name_entry_st *)') from a function with result type 'ASN1_OBJECT *(*)(X509_NAME_ENTRY *)' (aka 'struct asn1_object_st *(*)(struct X509_name_entry_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509_NAME_ENTRY_get_object;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3031:24: error: incompatible pointer types returning 'X509_EXTENSION *(CONF *, X509V3_CTX *, int, const char *)' (aka 'struct X509_extension_st *(struct conf_st *, struct v3_ext_ctx *, int, const char *)') from a function with result type 'X509_EXTENSION *(*)(CONF *, X509V3_CTX *, int, char *)' (aka 'struct X509_extension_st *(*)(struct conf_st *, struct v3_ext_ctx *, int, char *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509V3_EXT_nconf_nid;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3036:24: error: incompatible pointer types returning 'X509_EXTENSION *(CONF *, X509V3_CTX *, const char *, const char *)' (aka 'struct X509_extension_st *(struct conf_st *, struct v3_ext_ctx *, const char *, const char *)') from a function with result type 'X509_EXTENSION *(*)(CONF *, X509V3_CTX *, char *, char *)' (aka 'struct X509_extension_st *(*)(struct conf_st *, struct v3_ext_ctx *, char *, char *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return X509V3_EXT_nconf;
cargo:warning=                       ^~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3041:24: error: incompatible pointer types returning 'int (unsigned char **, const ASN1_STRING *)' (aka 'int (unsigned char **, const struct asn1_string_st *)') from a function with result type 'int (*)(unsigned char **, ASN1_STRING *)' (aka 'int (*)(unsigned char **, struct asn1_string_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return ASN1_STRING_to_UTF8;
cargo:warning=                       ^~~~~~~~~~~~~~~~~~~
cargo:warning=/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c:3066:24: error: incompatible pointer types returning 'int (const EVP_PKEY *)' (aka 'int (const struct evp_pkey_st *)') from a function with result type 'int (*)(EVP_PKEY *)' (aka 'int (*)(struct evp_pkey_st *)') [-Werror,-Wincompatible-pointer-types]
cargo:warning=                return EVP_PKEY_bits;
cargo:warning=                       ^~~~~~~~~~~~~
cargo:warning=fatal error: too many errors emitted, stopping now [-ferror-limit=]
cargo:warning=20 errors generated.
exit code: 1

--- stderr
thread 'main' panicked at '

Internal error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-o" "/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.o" "-c" "/rust/rust-openssl/target/debug/build/systest-8a0914cb68876f1c/out/all.c" with args "cc" did not execute successfully (status code exit code: 1).

', /rust/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.23/src/lib.rs:2253:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -76,9 +76,7 @@ macos_job: &MACOS_JOB
   - run: sudo chown -R $USER /usr/local /opt
   - run: curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain 1.20.0
   - run: sudo ln -s $CARGO_HOME/bin/* /usr/local/bin
-  - *RESTORE_REGISTRY
   - run: cargo generate-lockfile --verbose
-  - *SAVE_REGISTRY
   - run: echo "homebrew-x86_64-apple-darwin" > ~/lib_key
   - *RESTORE_DEPS
   - run: cargo run --manifest-path=systest/Cargo.toml --features "$FEATURES"

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -108,9 +108,9 @@ openssl_101: &OPENSSL_101
 libressl_250: &LIBRESSL_250
   LIBRARY: libressl
   VERSION: 2.5.0
-libressl_270: &LIBRESSL_272
+libressl_270: &LIBRESSL_280
   LIBRARY: libressl
-  VERSION: 2.7.2
+  VERSION: 2.8.0
 
 x86_64: &X86_64
   TARGET: x86_64-unknown-linux-gnu
@@ -195,10 +195,10 @@ jobs:
     <<: *JOB
     environment:
       <<: [*LIBRESSL_250, *X86_64, *BASE]
-  x86_64-libressl-2.7.2:
+  x86_64-libressl-2.8.0:
     <<: *JOB
     environment:
-      <<: [*LIBRESSL_272, *X86_64, *BASE]
+      <<: [*LIBRESSL_280, *X86_64, *BASE]
   macos:
     <<: *MACOS_JOB
     environment:
@@ -226,6 +226,6 @@ workflows:
     - armhf-openssl-1.1.0
     - armhf-openssl-1.0.2
     - x86_64-libressl-2.5.0
-    - x86_64-libressl-2.7.2
+    - x86_64-libressl-2.8.0
     - macos
     - macos-vendored

--- a/openssl-sys/build/cfgs.rs
+++ b/openssl-sys/build/cfgs.rs
@@ -16,6 +16,9 @@ pub fn get(openssl_version: Option<u64>, libressl_version: Option<u64>) -> Vec<&
         if libressl_version >= 0x2_07_03_00_0 {
             cfgs.push("libressl273");
         }
+        if libressl_version >= 0x2_08_00_00_0 {
+            cfgs.push("libressl280");
+        }
     } else {
         let openssl_version = openssl_version.unwrap();
 

--- a/openssl-sys/build/main.rs
+++ b/openssl-sys/build/main.rs
@@ -497,6 +497,7 @@ See rust-openssl README for more information:
             (6, 2) => ('6', '2'),
             (6, _) => ('6', 'x'),
             (7, _) => ('7', 'x'),
+            (8, 0) => ('8', 'x'),
             _ => version_error(),
         };
 

--- a/openssl-sys/src/asn1.rs
+++ b/openssl-sys/src/asn1.rs
@@ -49,7 +49,7 @@ extern "C" {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn ASN1_STRING_to_UTF8(out: *mut *mut c_uchar, s: *const ASN1_STRING) -> c_int;
         }

--- a/openssl-sys/src/bio.rs
+++ b/openssl-sys/src/bio.rs
@@ -68,6 +68,7 @@ cfg_if! {
     } else {
         extern "C" {
             pub fn BIO_s_file() -> *mut BIO_METHOD;
+            pub fn BIO_new(type_: *mut BIO_METHOD) -> *mut BIO;
         }
     }
 }

--- a/openssl-sys/src/bio.rs
+++ b/openssl-sys/src/bio.rs
@@ -36,7 +36,7 @@ pub type bio_info_cb =
     Option<unsafe extern "C" fn(*mut BIO, c_int, *const c_char, c_int, c_long, c_long)>;
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum BIO_METHOD {}
     } else {
         #[repr(C)]
@@ -60,24 +60,14 @@ pub unsafe fn BIO_get_mem_data(b: *mut BIO, pp: *mut *mut c_char) -> c_long {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn BIO_s_file() -> *const BIO_METHOD;
-        }
-    } else {
-        extern "C" {
-            pub fn BIO_s_file() -> *mut BIO_METHOD;
-        }
-    }
-}
-cfg_if! {
-    if #[cfg(ossl110)] {
-        extern "C" {
             pub fn BIO_new(type_: *const BIO_METHOD) -> *mut BIO;
         }
     } else {
         extern "C" {
-            pub fn BIO_new(type_: *mut BIO_METHOD) -> *mut BIO;
+            pub fn BIO_s_file() -> *mut BIO_METHOD;
         }
     }
 }
@@ -96,7 +86,7 @@ extern "C" {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn BIO_s_mem() -> *const BIO_METHOD;
         }
@@ -107,7 +97,7 @@ cfg_if! {
     }
 }
 cfg_if! {
-    if #[cfg(ossl102)] {
+    if #[cfg(any(ossl102, libressl280))] {
         extern "C" {
             pub fn BIO_new_mem_buf(buf: *const c_void, len: c_int) -> *mut BIO;
         }

--- a/openssl-sys/src/ec.rs
+++ b/openssl-sys/src/ec.rs
@@ -166,7 +166,7 @@ extern "C" {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum ECDSA_SIG {}
     } else {
         #[repr(C)]

--- a/openssl-sys/src/evp.rs
+++ b/openssl-sys/src/evp.rs
@@ -108,19 +108,28 @@ extern "C" {
         e: *mut ENGINE,
         pkey: *mut EVP_PKEY,
     ) -> c_int;
-    #[cfg(not(ossl102))]
-    pub fn EVP_DigestVerifyFinal(
-        ctx: *mut EVP_MD_CTX,
-        sigret: *mut c_uchar,
-        siglen: size_t,
-    ) -> c_int;
-    #[cfg(ossl102)]
-    pub fn EVP_DigestVerifyFinal(
-        ctx: *mut EVP_MD_CTX,
-        sigret: *const c_uchar,
-        siglen: size_t,
-    ) -> c_int;
+}
+cfg_if! {
+    if #[cfg(any(ossl102, libressl280))] {
+        extern "C" {
+            pub fn EVP_DigestVerifyFinal(
+                ctx: *mut EVP_MD_CTX,
+                sigret: *const c_uchar,
+                siglen: size_t,
+            ) -> c_int;
+        }
+    } else {
+        extern "C" {
+            pub fn EVP_DigestVerifyFinal(
+                ctx: *mut EVP_MD_CTX,
+                sigret: *mut c_uchar,
+                siglen: size_t,
+            ) -> c_int;
+        }
+    }
+}
 
+extern "C" {
     pub fn EVP_CIPHER_CTX_new() -> *mut EVP_CIPHER_CTX;
     pub fn EVP_CIPHER_CTX_free(ctx: *mut EVP_CIPHER_CTX);
     pub fn EVP_MD_CTX_copy_ex(dst: *mut EVP_MD_CTX, src: *const EVP_MD_CTX) -> c_int;
@@ -181,7 +190,7 @@ extern "C" {
     pub fn EVP_PKEY_id(pkey: *const EVP_PKEY) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn EVP_PKEY_bits(key: *const EVP_PKEY) -> c_int;
         }

--- a/openssl-sys/src/ossl_typ.rs
+++ b/openssl-sys/src/ossl_typ.rs
@@ -13,7 +13,7 @@ pub enum ASN1_OBJECT {}
 
 pub enum bio_st {} // FIXME remove
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum BIO {}
     } else {
         #[repr(C)]
@@ -59,7 +59,7 @@ pub enum BN_CTX {}
 pub enum BN_GENCB {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum EVP_CIPHER {}
     } else {
         #[repr(C)]
@@ -90,7 +90,7 @@ cfg_if! {
 pub enum EVP_CIPHER_CTX {}
 pub enum EVP_MD {}
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum EVP_MD_CTX {}
     } else {
         #[repr(C)]
@@ -105,7 +105,7 @@ cfg_if! {
     }
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum EVP_PKEY {}
     } else {
         #[repr(C)]
@@ -127,7 +127,7 @@ pub enum EVP_PKEY_ASN1_METHOD {}
 pub enum EVP_PKEY_CTX {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum HMAC_CTX {}
     } else {
         #[repr(C)]
@@ -143,7 +143,7 @@ cfg_if! {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum DH {}
     } else {
         #[repr(C)]
@@ -172,7 +172,7 @@ cfg_if! {
 pub enum DH_METHOD {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum DSA {}
     } else {
         #[repr(C)]
@@ -201,7 +201,7 @@ cfg_if! {
 pub enum DSA_METHOD {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum RSA {}
     } else if #[cfg(libressl)] {
         #[repr(C)]
@@ -267,7 +267,7 @@ pub enum RSA_METHOD {}
 pub enum EC_KEY {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum X509 {}
     } else if #[cfg(libressl)] {
         #[repr(C)]
@@ -344,7 +344,7 @@ pub enum X509_STORE {}
 pub enum X509_STORE_CTX {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum X509_VERIFY_PARAM {} 
     } else if #[cfg(libressl251)] {
         #[repr(C)]
@@ -406,7 +406,7 @@ pub enum OPENSSL_INIT_SETTINGS {}
 
 pub enum ENGINE {}
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum SSL {}
     } else if #[cfg(libressl251)] {
         #[repr(C)]
@@ -708,7 +708,7 @@ cfg_if! {
     }
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum SSL_CTX {}
     } else if #[cfg(libressl251)] {
         #[repr(C)]
@@ -970,7 +970,7 @@ cfg_if! {
 pub enum COMP_METHOD {}
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum CRYPTO_EX_DATA {}
     } else if #[cfg(libressl)] {
         #[repr(C)]

--- a/openssl-sys/src/pkcs12.rs
+++ b/openssl-sys/src/pkcs12.rs
@@ -18,7 +18,7 @@ extern "C" {
     ) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn PKCS12_create(
                 pass: *const c_char,

--- a/openssl-sys/src/ssl.rs
+++ b/openssl-sys/src/ssl.rs
@@ -25,7 +25,7 @@ pub const SSL_FILETYPE_ASN1: c_int = X509_FILETYPE_ASN1;
 pub enum SSL_METHOD {}
 pub enum SSL_CIPHER {}
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         pub enum SSL_SESSION {}
     } else if #[cfg(libressl251)] {
         #[repr(C)]
@@ -292,7 +292,13 @@ cfg_if! {
 pub const SSL_OP_ENABLE_MIDDLEBOX_COMPAT: c_ulong = 0x00100000;
 
 pub const SSL_OP_CIPHER_SERVER_PREFERENCE: c_ulong = 0x00400000;
-pub const SSL_OP_TLS_ROLLBACK_BUG: c_ulong = 0x00800000;
+cfg_if! {
+    if #[cfg(libressl280)] {
+        pub const SSL_OP_TLS_ROLLBACK_BUG: c_ulong = 0;
+    } else {
+        pub const SSL_OP_TLS_ROLLBACK_BUG: c_ulong = 0x00800000;
+    }
+}
 
 
 cfg_if! {
@@ -466,7 +472,7 @@ extern "C" {
     );
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn SSL_CTX_sess_set_get_cb(
                 ctx: *mut ::SSL_CTX,
@@ -497,7 +503,7 @@ extern "C" {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn SSL_CTX_set_cookie_verify_cb(
                 s: *mut SSL_CTX,
@@ -843,7 +849,7 @@ extern "C" {
     pub fn SSL_CIPHER_get_bits(cipher: *const SSL_CIPHER, alg_bits: *mut c_int) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn SSL_CIPHER_get_version(cipher: *const SSL_CIPHER) -> *const c_char;
         }
@@ -1036,11 +1042,20 @@ extern "C" {
     ) -> *mut c_char;
 
     pub fn SSL_get_certificate(ssl: *const SSL) -> *mut X509;
-    #[cfg(not(ossl102))]
-    pub fn SSL_get_privatekey(ssl: *mut SSL) -> *mut EVP_PKEY;
-    #[cfg(ossl102)]
-    pub fn SSL_get_privatekey(ssl: *const SSL) -> *mut EVP_PKEY;
+}
+cfg_if! {
+    if #[cfg(any(ossl102, libressl280))] {
+        extern "C" {
+            pub fn SSL_get_privatekey(ssl: *const SSL) -> *mut EVP_PKEY;
+        }
+    } else {
+        extern "C" {
+            pub fn SSL_get_privatekey(ssl: *mut SSL) -> *mut EVP_PKEY;
+        }
+    }
+}
 
+extern "C" {
     #[cfg(ossl102)]
     pub fn SSL_CTX_get0_certificate(ctx: *const SSL_CTX) -> *mut X509;
     #[cfg(ossl102)]

--- a/openssl-sys/src/x509.rs
+++ b/openssl-sys/src/x509.rs
@@ -191,7 +191,7 @@ extern "C" {
     pub fn X509_set_issuer_name(x: *mut X509, name: *mut X509_NAME) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509_get_issuer_name(x: *const ::X509) -> *mut ::X509_NAME;
         }
@@ -205,15 +205,24 @@ extern "C" {
     pub fn X509_set_subject_name(x: *mut X509, name: *mut X509_NAME) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509_get_subject_name(x: *const ::X509) -> *mut ::X509_NAME;
+        }
+    } else {
+        extern "C" {
+            pub fn X509_get_subject_name(x: *mut ::X509) -> *mut ::X509_NAME;
+        }
+    }
+}
+cfg_if! {
+    if #[cfg(ossl110)] {
+        extern "C" {
             pub fn X509_set1_notBefore(x: *mut ::X509, tm: *const ::ASN1_TIME) -> c_int;
             pub fn X509_set1_notAfter(x: *mut ::X509, tm: *const ::ASN1_TIME) -> c_int;
         }
     } else {
         extern "C" {
-            pub fn X509_get_subject_name(x: *mut ::X509) -> *mut ::X509_NAME;
             pub fn X509_set_notBefore(x: *mut ::X509, tm: *const ::ASN1_TIME) -> c_int;
             pub fn X509_set_notAfter(x: *mut ::X509, tm: *const ::ASN1_TIME) -> c_int;
         }
@@ -244,7 +253,7 @@ extern "C" {
 }
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509_NAME_entry_count(n: *const X509_NAME) -> c_int;
         }
@@ -255,11 +264,19 @@ cfg_if! {
     }
 }
 
-extern "C" {
-    pub fn X509_NAME_get_index_by_NID(n: *mut X509_NAME, nid: c_int, last_pos: c_int) -> c_int;
+cfg_if! {
+    if #[cfg(libressl280)] {
+        extern "C" {
+            pub fn X509_NAME_get_index_by_NID(n: *const X509_NAME, nid: c_int, last_pos: c_int) -> c_int;
+        }
+    } else {
+        extern "C" {
+            pub fn X509_NAME_get_index_by_NID(n: *mut X509_NAME, nid: c_int, last_pos: c_int) -> c_int;
+        }
+    }
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509_NAME_get_entry(n: *const X509_NAME, loc: c_int) -> *mut X509_NAME_ENTRY;
             pub fn X509_NAME_add_entry_by_NID(
@@ -305,7 +322,7 @@ extern "C" {
     pub fn X509_add_ext(x: *mut X509, ext: *mut X509_EXTENSION, loc: c_int) -> c_int;
 }
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509_get_ext_d2i(
                 x: *const ::X509,

--- a/openssl-sys/src/x509v3.rs
+++ b/openssl-sys/src/x509v3.rs
@@ -41,7 +41,7 @@ pub const X509_CHECK_FLAG_SINGLE_LABEL_SUBDOMAINS: c_uint = 0x10;
 pub const X509_CHECK_FLAG_NEVER_CHECK_SUBJECT: c_uint = 0x20;
 
 cfg_if! {
-    if #[cfg(ossl110)] {
+    if #[cfg(any(ossl110, libressl280))] {
         extern "C" {
             pub fn X509V3_EXT_nconf_nid(
                 conf: *mut CONF,

--- a/openssl/build.rs
+++ b/openssl/build.rs
@@ -49,5 +49,9 @@ fn main() {
         if version >= 0x2_07_03_00_0 {
             println!("cargo:rustc-cfg=libressl273");
         }
+
+        if version >= 0x2_08_00_00_0 {
+            println!("cargo:rustc-cfg=libressl280");
+        }
     }
 }

--- a/openssl/src/ssl/callbacks.rs
+++ b/openssl/src/ssl/callbacks.rs
@@ -379,10 +379,13 @@ pub unsafe extern "C" fn raw_remove_session<F>(
     callback(ctx, session)
 }
 
-#[cfg(ossl110)]
-type DataPtr = *const c_uchar;
-#[cfg(not(ossl110))]
-type DataPtr = *mut c_uchar;
+cfg_if! {
+    if #[cfg(any(ossl110, libressl280))] {
+        type DataPtr = *const c_uchar;
+    } else {
+        type DataPtr = *mut c_uchar;
+    }
+}
 
 pub unsafe extern "C" fn raw_get_session<F>(
     ssl: *mut ffi::SSL,
@@ -503,11 +506,13 @@ where
     }
 }
 
-#[cfg(ossl110)]
-type CookiePtr = *const c_uchar;
-
-#[cfg(not(ossl110))]
-type CookiePtr = *mut c_uchar;
+cfg_if! {
+    if #[cfg(any(ossl110, libressl280))] {
+        type CookiePtr = *const c_uchar;
+    } else {
+        type CookiePtr = *mut c_uchar;
+    }
+}
 
 pub extern "C" fn raw_cookie_verify<F>(
     ssl: *mut ffi::SSL,

